### PR TITLE
Add play-until functionality

### PR DESCRIPTION
**Pull request by gbiggs**
This PR adds the ability to play until a given timestamp.

This work was co-authored by @agalbachicar.

This PR requires #960 to be merged first (at which point it will probably need to be rebased).

### CODE REVIEW

**Comment by MichaelOrlov** [comment]
@gbiggs Let's finish review for #960 first.

File: rosbag2_py/src/rosbag2_py/_transport.cpp [hunk 1]
```diff
@@ -275,7 +297,12 @@ PYBIND11_MODULE(_transport, m) {
 
   py::class_<rosbag2_py::Player>(m, "Player")
   .def(py::init())
-  .def("play", &rosbag2_py::Player::play)
+  .def(
+    "play", &rosbag2_py::Player::play, py::arg("storage_options"), py::arg(
+      "play_options"), py::arg("duration") = std::nullopt)
+  .def(
+    "play_until", &rosbag2_py::Player::play, py::arg("storage_options"), py::arg(
```
Comment (hidmic) [inline rosbag2_py/src/rosbag2_py/_transport.cpp:304]: @gbiggs should be `rosbag2_py::Player::play_until` instead?

File: rosbag2_py/src/rosbag2_py/_transport.cpp [hunk 2]
```diff
@@ -275,7 +297,12 @@ PYBIND11_MODULE(_transport, m) {
...
+      "play_options"), py::arg("timestamp") = std::nullopt)
```
Comment (hidmic) [inline rosbag2_py/src/rosbag2_py/_transport.cpp:305]: @gbiggs hmm, shouldn't `timestamp` be mandatory for `play_until()`?

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 3]
```diff
@@ -658,6 +690,34 @@ void Player::create_control_services()
...
+    "~/play_until",
+    [this](
+      const std::shared_ptr<rmw_request_id_t>/* request_header */,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayUntil::Request> request,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayUntil::Response> response)
+    {
+      const rcutils_time_point_value_t timestamp =
+      static_cast<rcutils_time_point_value_t>(request->time.sec) *
+      static_cast<rcutils_time_point_value_t>(1000000000) +
+      static_cast<rcutils_time_point_value_t>(request->time.nanosec);
```
Comment (hidmic) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:717]: @gbiggs nit: consider
```suggestion
      const rcutils_time_point_value_t timestamp =
          rclcpp::Time(request->time).nanoseconds();
```

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    auto primitive_message1 = get_messages_basic_types()[0];
+    primitive_message1->int32_value = kIntValue;
+
+    auto complex_message1 = get_messages_arrays()[0];
+    complex_message1->float32_values = {{kFloat1Value, kFloat2Value, kFloat3Value}};
+    complex_message1->bool_values = {{kBool1Value, kBool2Value, kBool3Value}};
+
+    // @{ Ordering matters. The mock reader implementation moves messages
+    //    around without any knowledge about message chronology. It just picks
+    //    the next one Make sure to keep the list in order or sort it!
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:104]: @gbiggs nit: 
```suggestion
    //    the next one. Make sure to keep the list in order or sort it!
```

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+  {
+    auto topic_types = get_topic_types();
+    auto messages = get_serialized_messages();
+
+    auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+    prepared_mock_reader->prepare(messages, topic_types);
+    auto reader = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+
+    // Due to a problem related to the subscriber, we play many (3) messages but make the subscriber
+    // node spin only until 2 have arrived. Hence the 2 as `launch_subscriber()` argument.
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:126]: @gbiggs can we be a bit more specific? Is there a tracking issue we can link here?

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    sub_->add_subscription<test_msgs::msg::BasicTypes>(kTopic1, 2);
+    sub_->add_subscription<test_msgs::msg::Arrays>(kTopic2, 2);
+
+    player_ = std::make_shared<rosbag2_transport::Player>(
+      std::move(
+        reader), storage_options_, play_options_);
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:132]: @gbiggs nit: can we collapse this two lines into one?

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+  std::shared_ptr<rosbag2_transport::Player> player_;
+  std::future<void> await_received_messages_;
+};
+
+TEST_F(RosBag2PlayForTestFixture, play_for_all_are_played_due_to_duration)
+{
+  auto await_received_messages = sub_->spin_subscriptions();
+
+  player_->play(std::chrono::nanoseconds(std::chrono::milliseconds(1000)).count());
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:143]: @gbiggs fyi: you could also do
```suggestion
  player_->play(RCUTILS_MS_TO_NS(1000));
```
or equivalently
```suggestion
  player_->play(RCUTILS_S_TO_NS(1));
```
Same everywhere else.

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+  await_received_messages.get();
+
+  auto replayed_test_primitives = sub_->get_received_messages<test_msgs::msg::BasicTypes>(
+    kTopic1);
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:148]: @gbiggs nit: 
```suggestion
  auto replayed_test_primitives = 
      sub_->get_received_messages<test_msgs::msg::BasicTypes>(kTopic1);
```
Same elsewhere.

File: rosbag2_transport/test/rosbag2_transport/test_play_for.cpp [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    player_ = std::make_shared<rosbag2_transport::Player>(
+      std::move(
+        reader), storage_options_, play_options_);
+  }
+
+  std::shared_ptr<rosbag2_transport::Player> player_;
+  std::future<void> await_received_messages_;
+};
+
+TEST_F(RosBag2PlayForTestFixture, play_for_all_are_played_due_to_duration)
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_for.cpp:139]: @gbiggs nit^N: test case names read a bit awkward. For instance, how about `play_for_a_long_duration_publishes_some` here?

File: rosbag2_transport/test/rosbag2_transport/test_play_until.cpp [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+  await_received_messages.get();
+
+  auto replayed_test_primitives =
+    sub_->get_received_messages<test_msgs::msg::BasicTypes>("/topic1");
+  // No messages are allowed to have arrived
+  EXPECT_THAT(replayed_test_primitives, SizeIs(0u));
+
+  auto replayed_test_arrays = sub_->get_received_messages<test_msgs::msg::Arrays>("/topic2");
+  // All messages should have arrived.
```
Comment (hidmic) [inline rosbag2_transport/test/rosbag2_transport/test_play_until.cpp:328]: @gbiggs leftover?

**Review by hidmic — COMMENTED** [review]
First pass!

**Comment by agalbachicar** [comment]
Before proceeding to fix this PR, I'd like to know your preference on the implementation @MichaelOrlov and @hidmic . We proposed and discussed this in the [meeting group on 2021-05-21](https://docs.google.com/document/d/1UXWaEkT6ji0Z0uTtJJSni5Y6fiopM9qOOzv4ldeNaAM/edit?pli=1#heading=h.lazc4q6xqmm6). The feature request changed a bit to be: "play until timepoint" rather than "play until next message of topic T" since then.

Also, we needed to have this functionality in another rosbag2 fork first that should  be compliant with foxy. As a consequence, this patch looks so odd at first (likewise to #960 at first). We aimed for API compatibility or minimal changes across ros2 releases. That said, we are eager to push to get these changes in before the code freeze (as everyone else contributing :)) and hoping you could make some time to guide us to get these changes in.

Looking forward to hearing your thoughts.

**Comment by MichaelOrlov** [comment]
@agalbachicar @gbiggs I think you will need to rebase first and rework this PR since we already have `play_messages_from_queue(play_until_time);` API implemented in #960.
Implementation for this PR is going to be sort of a light-way.
I assume will need to add new parameter for command line and in playback options data structure. Further in `Player::Play()` use two parameters `play_for` and `play_until` to determine which time is latter, to use it as parameter for existing `play_messages_from_queue(play_until_time);` call.
Please take in to consideration the way how we workout unit tests in #960 and use similar approach.

**Comment by gbiggs** [comment]
Closing in favour of #1005

**Pull request closed without merging**